### ISSUE TITLE
A few thoughts about p5.Vector implementation

### ISSUE THREAD

**Issue by tawez** [processing/p5.js#2318]
I can make PR, but first I would like to discuss the list below.

1.  Whenever Vector instance is created with `new p5.Vector(...)`, the default value
    for x, y and z is 0.

    Then in every Vector method that accepts Vector as argument value, default 0 is applied again and again, for example:
    ```javascript
    p5.Vector.prototype.set = function (x, y, z) {
      if (x instanceof p5.Vector) {
        this.x = x.x || 0;
        this.y = x.y || 0;
        this.z = x.z || 0;
        return this;
      }
      ...
    };
    ```
    I think it is not necessary, because the default value is already there.

2.  The default value of `n` in `p5.Vector.prototype.mult` is 0.
    When I call `mult` with no argument, my vector will be mutated.
    I think the default value of `n` should be 1.

3.  We don't check if we divide by 0 in `p5.Vector.prototype.div`.
    There is a lot of validation in other methods, but not in `div`.
    Is it done by design?

    Checking divide by 0 case in `div` method would make `p5.Vector.prototype.normalize` a bit simpler.

4.  Performance optimisations:

    - this.p5.random() => Math.random()
    - Math.PI*2 => constants.TWO_PI

    BTW. I can't get the fundamental difference between `if` and `else` parts,
    as the results seems to be the same, except the `if` part will be slower.

    ```javascript
    p5.Vector.random3D = function random3D() {
      ...
      if (this.p5) {
        angle = this.p5.random(0,constants.TWO_PI);
        vz = this.p5.random(-1,1);
      } else {
        angle = Math.random()*Math.PI*2;
        vz = Math.random()*2-1;
      }
      ...
    }
    ```

5.  Constructor could be much simpler if there were named arguments and different order of them.
    `function() => function(x, y, z, context)`

    ```javascript
    p5.Vector = function(x, y, z, context) {
      // This is how it comes in with createVector()
      if (context instanceof p5) {
        this.p5 = context;
      }
      this.x = x || 0;
      this.y = y || 0;
      this.z = z || 0;
      this.name = 'p5.Vector';
    };
    ```
    This of course should be propagated in other places, but there is only a few such a cases. 

6.  Functions should be named. This will help during debugging.

    ```javascript
    // Bad example
    p5.Vector.prototype.set = function (x, y, z) {...}

    // Good example
    p5.Vector.prototype.set = function set(x, y, z) {...}

    // Still good, but over engineered example
    p5.Vector.prototype.set = function p5VectorSet(x, y, z) {...}
    ```

**Comment by Jared-Sprague** [comment]
Hey!  Some great suggestions!

Here is my feedback:

>1.  Whenever Vector instance is created with new p5.Vector(...), the default value
for x, y and z is 0.
>
>Then in every Vector method that accepts Vector as argument value, default 0 is applied again and again, for example:

I think it's necessary to default to 0 again because you can't assume that the instance of p5.Vector being passed in has valid values for x, y, z.

> 2. The default value of n in p5.Vector.prototype.mult is 0.
> When I call mult with no argument, my vector will be mutated.
> I think the default value of n should be 1.

Agreed, better yet, validate that n defined and a number otherwise return and not do anything.  e.g.
```JavaScript
p5.Vector.prototype.mult = function (n) {
  if (Number.isNaN(n)) return;
  this.x *= n;
  this.y *= n;
  this.z *= n;
  return this;
};
```

> 3. We don't check if we divide by 0 in p5.Vector.prototype.div.
> There is a lot of validation in other methods, but not in div.
> Is it done by design?

I think that adding validation for divide by 0 is unnecessary, since JavaScript will return Infinity for divide by 0,  I say just let JavaScript handle it.  However,  I definitely think that that n should have the same validation as mentioned above for mult.

> Performance optimisations:

Totally agree with these.

> 4. Constructor could be much simpler if there were named arguments and different order of them.
> function() => function(x, y, z, context)

This would be an api breaking change.  Any applications using passing an instance of p5 as the first argument would stop working.  IMO, api breaking changes shouldn't be made lightly, and should be made backwards compatible if possible.

> 5. Functions should be named. This will help during debugging.

Totally agree,  I even like the longer format as it makes it very clear what function you're looking at when debugging.

**Comment by sixhat** [comment]
I totally agree with @Jared-Sprague comments. With the computing power available these days code legibility should trump all other considerations, except correctness. 

The only point I have some doubts is n. 3 because letting JS divide by 0 and return **Infinity** might not be detected by the user (it doesn't give an error, just returns Infinity). There might even exist some code that relies on this value to perform some computations.

I think that the minimum we could do was to check for div by 0 and write a warning to console. This way the use would be notified of a div by 0 while debugging the program.

**Comment by tawez** [comment]
@sixhat The computational power is an illusory when we talk about core functions. ;)
The computational overhead is accumulating very fast.

1.  Considering for example implementation of `div` method and the overall uncertainty about how `Vector` is used I agree not to change anything.
2.  @Jared-Sprague I've presented the problem from mathematical point of view.
    Yes, this is more or less the way I was going to implement it ;)
    but I was thinking about `isFinite()` method.
3.  @sixhat I would prefer to have Vector consistent from mathematical point of view.
    Everything else is just the implementation.
    (In general) I don't know if when someone is basing on bug in implementation is it still a bug or is it a feature? ;)
4. I see no objections so far
5.  As far as I was able to check, the second way the constructor is invoked (`new Vector(instanceof p5, Array)`)
    - is not documented;
    - is used internally by `createVector` and a few Vector methods;

    so for me it doesn't look like a breaking change.
    And so, this is why I am asking, cause I may not know about everything.
6.  I would prefer short version. Otherway we will end with JavaLikeMethodNamesThatAreLongerThanNecessary.
    With long names it would be a temptation to distinguish instance methods and static methods... we can never stop ;)
    I thing that error stack will give enough hints to find the source.

    We can start with short names and check how it works.
    Then we can change it when needed.

**Comment by Jared-Sprague** [comment]
Yeah, I agree at a minimum throwing a warning when attempting to divide by 0.

> As far as I was able to check, the second way the constructor is invoked (new Vector(instanceof p5, Array))
> 
> is not documented;
> is used internally by createVector and a few Vector methods;
> so for me it doesn't look like a breaking change.
> And so, this is why I am asking, cause I may not know about everything.

Anytime you re-order the signature of a public method, is the definition of an api breaking change,  because you are moving the position of the p5 argument from the first argument in the signature to the last,  and you can't guarantee that no clients exists in the world that are expecting the current signature.  It doesn't matter if it's not documented.   So if you re-order the method signature of any public method you should:
    a. Have a very good reason and benefit, to justify possibly breaking users code.  (I'm not sold on the benefit)
    b. If gain is worth it then you should also attempt make it backward compatible.

Yeah agree, that long names can get out of control.  Any name is better than no name :)

**Comment by meiamsome** [comment]
Disagree with part 4 greatly, it is very important to use p5.random in all places where it is possible as some sketches could rely on the fact that this RNG can be seeded. 

The else part is only there in case someone creates a vector without a p5 object, which I guess is technically possible.

**Comment by sixhat** [comment]
@meiamsome Indeed, overlooked the random part in point 4. One always have to be careful with side effects of changes.

**Comment by tawez** [comment]
@Jared-Sprague IMHO the contract between the author and the user is the API (in case of p5 expressed in the documentation).
All undocumented features and internal implementations are the subject of change.
If someone is using undocumented features or internal implementation he or she is doing this on his or hers own risk.
If the change of internal implementation does not affect API - it's not a breaking change.


To sum up current discussion...
Planned scope of PR:

1. `p5.Vector.prototype.mult` will do nothing if `n` is undefined or not a valid number
2. `p5.Vector.prototype.div` will throw a console warning when attempting to divide by 0
3. Change `Math.PI*2` to `constants.TWO_PI`
4. Name all anonymous functions

Not so much but not too bad ;)

**Comment by Jared-Sprague** [comment]
@tawez Sounds good! 

I'll just add one thing:

> p5.Vector.prototype.div will throw a console warning when attempting to divide by 0

I think this should also include the same validation for `n` as mult.

**Comment by tawez** [comment]
I am implementing 1. and 2. and need some guidance, how to report in p5 an issue propperly?

`console.warn(...)` seems to be the most straightforward, but there is also `src/core/error_helpers.js` with a bunch of methods.

**Comment by lmccart** [comment]
@tawez what is the issue you're trying to report?

**Comment by lmccart** [comment]
@tawez thank you for this. I'm having a really hard time following all the different points in one issue thread. would you mind opening new issues for each point you're suggesting implementing (ie one for issue #1 and issue #2 and we can discuss from there?) thanks!
